packages/host/app/lib/gc-identity-context.ts

```diff
@@ -53,69 +105,88 @@ export default class IdentityContextWithGarbageCollection
 
   sweep() {
     let consumptionGraph = this.makeConsumptionGraph();
-    for (let { card: instance } of this.#cards.values()) {
+    let cache = new Map<string, boolean>();
+    let visited = new WeakSet<CardDef>();
+    for (let instance of this.#cards.values()) {
       if (!instance) {
         continue;
       }
-      if (this.gcVisit(instance.id, consumptionGraph)) {
-        if (this.#gcCandidates.has(instance.id)) {
-          console.log(`garbage collecting instance ${instance.id} from store`);
+      if (visited.has(instance)) {
+        continue;
+      }
+      visited.add(instance);
+      if (this.gcVisit(instance[this.#api.localId], consumptionGraph, cache)) {
+        if (this.#gcCandidates.has(instance[this.#api.localId])) {
+          console.log(
+            `garbage collecting instance ${instance[this.#api.localId]} (remote id: ${instance.id}) from store`,
+          );
           // brand the instance to make it easier for debugging
           (instance as unknown as any)[
             Symbol.for('__instance_detached_from_store')
           ] = true;
-          this.delete(instance.id);
+          this.delete(instance[this.#api.localId]);
+          if (instance.id) {
+            this.delete(instance.id);
+          }
         } else {
-          this.#gcCandidates.add(instance.id);
+          this.#gcCandidates.add(instance[this.#api.localId]);
         }
       } else {
-        this.#gcCandidates.delete(instance.id);
+        this.#gcCandidates.delete(instance[this.#api.localId]);
       }
     }
   }
 
   gcVisit(
-    id: string,
-    consumptionGraph: Map<string, string[]>,
-    hasSubscribers = new Map<string, boolean>(),
-  ): boolean {
-    let cached = hasSubscribers.get(id);
+    localId: string,
+    consumptionGraph: Map<string, Set<string>>,
+    cache: Map<string, boolean>,
+  ): boolean /* true = eligible for GC, false = not eligible for GC */ {
```

Comment:
Would changing the name of the function make this more clear, like `checkGcEligibility`

Reply:
sure--good suggestion